File: design/Low Level Design LLD/LRU.js

// https://leetcode.com/problems/lru-cache/

function ListNode(key, val) {
    this.key = key;
    this.val = val;
    this.prev = null;
    this.next = null;
}

/**
 * @param {number} capacity
 */
var LRUCache = function (capacity) {
    this.capacity = capacity;
    this.cache = new Map();

    // Dummy head and tail
    this.head = new ListNode(0, 0);
    this.tail = new ListNode(0, 0);

    this.head.next = this.tail;
    this.tail.prev = this.head;
};

/**
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function (key) {
    // Key not found
    if (!this.cache.has(key)) return -1;

    // Move accessed node to the front (Most Recently Used)
    const node = this.cache.get(key);
    this._remove(node);
    this._add(node);

    return node.val;
};

/**
 * @param {number} key
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function (key, value) {

    // If key already exists, remove old node
    if (this.cache.has(key)) {
        const node = this.cache.get(key);
        this._remove(node);
        this.cache.delete(key);
    }

    // Insert new node at the front
    const newNode = new ListNode(key, value);
    this._add(newNode);

    // Remove Least Recently Used node if capacity exceeded
    if (this.cache.size > this.capacity) {
        const lru = this.tail.prev;
        this._remove(lru);
        this.cache.delete(lru.key);
    }
};

/**
 * Add node right after head (Most Recently Used position)
 */
LRUCache.prototype._add = function (node) {
    const next = this.head.next;

    this.head.next = node;
    node.prev = this.head;

    node.next = next;
    next.prev = node;

    this.cache.set(node.key, node);
};

/**
 * Remove node from the doubly linked list
 */
LRUCache.prototype._remove = function (node) {
    const prev = node.prev;
    const next = node.next;

    prev.next = next;
    next.prev = prev;
};

/**
 * Your LRUCache object will be instantiated and called as such:
 *
 * const obj = new LRUCache(capacity);
 * const value = obj.get(key);
 * obj.put(key, value);
 */

File: design/Coding Problems/LRU.java
// https://leetcode.com/problems/lru-cache/

import java.util.HashMap;
import java.util.Map;

class LRUCache {

    private static class Node {
        int key;
        int val;
        Node prev;
        Node next;

        Node(int key, int val) {
            this.key = key;
            this.val = val;
        }
    }

    private final int capacity;
    private final Map<Integer, Node> cache;

    // Dummy head and tail
    private final Node head;
    private final Node tail;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.cache = new HashMap<>();

        this.head = new Node(0, 0);
        this.tail = new Node(0, 0);

        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        // Key not found
        if (!cache.containsKey(key)) return -1;

        // Move accessed node to the front (Most Recently Used)
        Node node = cache.get(key);
        remove(node);
        add(node);

        return node.val;
    }

    public void put(int key, int value) {

        // If key already exists, remove old node
        if (cache.containsKey(key)) {
            Node existing = cache.get(key);
            remove(existing);
            cache.remove(key);
        }

        // Insert new node at the front
        Node newNode = new Node(key, value);
        add(newNode);

        // Remove Least Recently Used node if capacity exceeded
        if (cache.size() > capacity) {
            Node lru = tail.prev;
            remove(lru);
            cache.remove(lru.key);
        }
    }

    // Add node right after head (Most Recently Used position)
    private void add(Node node) {
        Node next = head.next;

        head.next = node;
        node.prev = head;

        node.next = next;
        next.prev = node;

        cache.put(node.key, node);
    }

    // Remove node from the doubly linked list
    private void remove(Node node) {
        Node prev = node.prev;
        Node next = node.next;

        prev.next = next;
        next.prev = prev;
    }
}

/*
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int value = obj.get(key);
 * obj.put(key, value);
 */
